### PR TITLE
Migrate kubelet to ConfigMapManager interface and use TTL-based caching manager

### LINKED ISSUE #41379: Add a configmap manager to cache duplicate configmaps
Clone of https://github.com/kubernetes/kubernetes/issues/19188 for configmaps

A secret manager was added to the kubelet to reduce duplicate requests for secrets (if a secret was used in multiple volumes, pull secrets, and envvars)

The same should be done for configmaps to keep parity between those two objects.

cc @wojtek-t @yujuhong

### PR DESCRIPTION
**Pull request by shyamjvs**
Fixes #41379
Sometime ago we moved to a secret manager interface for kubelet to manage secrets.
This PR's first commit moves config map management also to a similar interface.
The second commit adds TTL-based CachingConfigMapManager (similar to CachingSecretManager) and makes kubelet use it.

/cc @kubernetes/sig-node-pr-reviews @kubernetes/sig-scalability-misc @wojtek-t @dchen1107

### CODE REVIEW

**Comment by xiangpengzhao** [comment]
@shyamjvs please fix the test flake. It's due to your modification of `NewBasicPodManager`

```
I0525 22:15:52.975] k8s.io/kubernetes/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:44: not enough arguments in call to pod.NewBasicPodManager
I0525 22:15:52.976] 	have (*"k8s.io/kubernetes/pkg/kubelet/pod/testing".FakeMirrorClient, secret.Manager)
I0525 22:15:52.977] 	want (pod.MirrorClient, secret.Manager, configmap.Manager)
```

File: pkg/kubelet/pod/pod_manager.go [hunk 1]
```diff
@@ -163,6 +166,10 @@ func (pm *basicManager) updatePodsInternal(pods ...*v1.Pod) {
 			// not register pod, as it doesn't really matter.
 			pm.secretManager.RegisterPod(pod)
 		}
+		if pm.configMapManager != nil {
+			// TODO: Same as above.
```
Comment (wojtek-t) [inline pkg/kubelet/pod/pod_manager.go]: Please copy a TODO instead of writing "Same as above"

**Review by wojtek-t — COMMENTED** [review]
Just minor nit - also please fix tests.

File: pkg/kubelet/pod/pod_manager.go (hunk 1 above)
Comment (shyamjvs) [inline pkg/kubelet/pod/pod_manager.go]: Done.

**Comment by wojtek-t** [comment]
LGTM

@dchen1107 @yujuhong - can you of you please take a look?

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-unit test this
(seems unrelated)

**Comment by shyamjvs** [comment]
@wojtek-t @yujuhong FYI, I've added a second commit for TTL-based caching configmap manager using the ConfigMapManager interface introduced in the first commit. Again, this is just a copy of the logic of CachingSecretManager.

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-unit test this
(seems unrelated)

**Comment by yujuhong** [comment]
> Again, this is just a copy of the logic of CachingSecretManager.

You were telling the truth....they are the same even down to the unit tests. I am going to approve the change as it is given that I reviewed the original change (#36984), and keeping the secret and configmap managers consistent is important. We'll leave improvements for further changes.

@wojtek-t, I'm leaving this to you to review and lgtm the second commit.

/approve

**Comment by wojtek-t** [comment]
@yujuhong - thanks a lot. Really appreciate given how close to code freeze we are.
I reviewed the whole code, and this is really the copy. So there shouldn't be any problems with it.

FYI - the proper watch-based solution is still under design (we are solving much more generic problem with it), but I hope we will have the design approved this quarter and I still hope we will be able to implement it in 1.8. That should be huge win from the qps-limit perspective for the kubelet.

/lgtm
/approve

**Comment by wojtek-t** [comment]
@shyamjvs - please rebase

**Comment by shyamjvs** [comment]
@yujuhong Thanks a lot for taking out time to review.
@wojtek-t Rebased. Thanks.

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this
Seems some issue on gce side:
```
W0531 01:56:45.127] ERROR: (gcloud.compute.networks.create) Could not fetch resource:
W0531 01:56:45.128]  - Code: '3768090691729186078'
```

**Comment by shyamjvs** [comment]
Applying lgtm based on @wojtek-t 's https://github.com/kubernetes/kubernetes/pull/46470#issuecomment-305094034 above.

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-verify test this

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by shyamjvs** [comment]
Added 'retest-not-required' label as it happened for the 3rd time that all PR tests passed when I rerun them, but 'pull-kubernetes-e2e-gce-etcd3' (which has been super-flaky recently) fails during the premerge round, making me repeat the cycle all over again.

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by spiffxp** [comment]
looks like etcd3 job is bumping into #46713

**Comment by shyamjvs** [comment]
@spiffxp Thanks for the lead. Will wait for the new run's result (after bumping up subnet quota in the project) and then retry.

**Comment by shyamjvs** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by wojtek-t** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**